rpc/wrtc_call_queue_mongodb.go

```diff
@@ -111,8 +277,532 @@ const (
+		queue.csStateMu.Lock()
+		defer queue.csStateMu.Unlock()
+		delete(exchangeSubs, exchange)
+		if len(exchangeSubs) == 0 {
+			delete(queue.callExchangeSubs, callID)
+		}
+	}
+}
+
+// subscribeForNewCallOnHosts allows an answerer to subscribe for new calls on any of the given hosts. It returns
+// once this operator's change stream is tracking the hosts. The channel will have an event on it once
+// any of the hosts receives a call and it is also routed to this subscriber.
+func (queue *mongoDBWebRTCCallQueue) subscribeForNewCallOnHosts(
+	ctx context.Context,
+	hosts []string,
+) (<-chan mongodbCallEvent, func(), error) {
+	queue.csStateMu.Lock()
+	subChan := make(chan mongodbCallEvent, 1)
+	ready := make(chan struct{})
+	callEventHandler := &mongodbNewCallEventHandler{
+		eventChan: subChan,
+	}
+
+	var alreadyTrackedCount int
+	for _, host := range hosts {
+		if _, ok := queue.csTrackingHosts[host]; ok {
+			alreadyTrackedCount++
+		}
+		hostSubs, ok := queue.waitingForNewCallSubs[host]
+		if !ok {
+			hostSubs = map[*mongodbNewCallEventHandler]struct{}{}
+			queue.waitingForNewCallSubs[host] = hostSubs
+		}
+		hostSubs[callEventHandler] = struct{}{}
+	}
+
+	unsub := func() {
+		queue.csStateMu.Lock()
+		defer queue.csStateMu.Unlock()
+		for _, host := range hosts {
+			delete(queue.waitingForNewCallSubs[host], callEventHandler)
+			if len(queue.waitingForNewCallSubs[host]) == 0 {
+				delete(queue.waitingForNewCallSubs, host)
+			}
+		}
+	}
+
+	if alreadyTrackedCount == len(hosts) {
+		queue.csStateMu.Unlock()
+		return subChan, unsub, nil
+	}
+
+	queue.csAnswerersWaitingForNextCS = append(queue.csAnswerersWaitingForNextCS, func() {
+		close(ready)
+	})
+	queue.csStateMu.Unlock()
+	queue.csManagerSeq.Add(1)
+
+	select {
+	case <-ctx.Done():
+		unsub()
+		return nil, nil, ctx.Err()
+	case <-ready:
+		return subChan, unsub, nil
+	}
+}
+
+var (
+	projectStage   = bson.D{{"$project", bson.D{{webrtcOperatorHostsField, 1}, {"_id", 0}}}}
+	unwindAggStage = bson.D{{"$unwind", "$" + webrtcOperatorHostsField}}
+	groupAggStage  = bson.D{{"$group", bson.D{
+		{"_id", "$" + webrtcOperatorHostsHostCombinedField},
+		{"caller_size", bson.D{{"$sum", "$" + webrtcOperatorHostsCallerSizeCombinedField}}},
+		{"answerer_size", bson.D{{"$sum", "$" + webrtcOperatorHostsAnswererSizeCombinedField}}},
+	}}}
+)
+
+var errTooManyConns = status.Error(codes.Unavailable, "too many connection attempts; please wait a bit try again")
+
+func (queue *mongoDBWebRTCCallQueue) checkHostQueueSize(ctx context.Context, forCaller bool, hosts ...string) error {
+	hostsMatch := bson.D{
+		{"$match", bson.D{{webrtcOperatorHostsHostCombinedField, bson.D{{"$in", hosts}}}}},
+	}
+	pipeline := []interface{}{
+		hostsMatch,
+		projectStage,
+		unwindAggStage,
+		hostsMatch,
+		groupAggStage,
+	}
+	if forCaller {
+		pipeline = append(pipeline, queue.hostCallerQueueSizeMatchAggStage)
+	} else {
+		pipeline = append(pipeline, queue.hostAnswererQueueSizeMatchAggStage)
+	}
```

Comment:
[opt] This is cool how we are using aggregations todo the logical comparison. But it took me a while to wrap my head around it. Might be more readable if we just do the comparison in go?
`if (forCaller && ret[0]['caller_size'] < maxHostCallers ){`
...

Reply:
that puts more burden on us and increases the chance of fetching documents that do not need to be returned to us (network + compute costs), so I think it's better to keep in the DB.